ivy/functional/frontends/numpy/func_wrapper.py

```diff
@@ -0,0 +1,95 @@
+import ivy
+import functools
+from typing import Callable
+
+
+def _is_same_kind_or_safe(t1, t2):
+    if ivy.is_float_dtype(t1):
+        return ivy.is_float_dtype(t2) or ivy.can_cast(t1, t2)
+    elif ivy.is_uint_dtype(t1):
+        return ivy.is_uint_dtype(t2) or ivy.can_cast(t1, t2)
+    elif ivy.is_int_dtype(t1):
+        return ivy.is_int_dtype(t2) or ivy.can_cast(t1, t2)
+    elif ivy.is_bool_dtype(t1):
+        return ivy.is_bool_dtype(t2) or ivy.can_cast(t1, t2)
+    raise ivy.exceptions.IvyException(
+        "dtypes of input must be float, int, uint, or bool"
+    )
+
+
+def _assert_args_and_fn(args, dtype, fn):
+    ivy.assertions.check_all_or_any_fn(
+        *args,
+        fn=fn,
+        type="all",
+        message="type of input is incompatible with dtype: {}".format(dtype),
+    )
+
+
+def handle_numpy_casting(num_args: int = 1) -> Callable:
```

Comment:
see comment below, passing in `num_args` should not be necessary once we enable dynamic inference of the number of arrays and their location in the input arguments and keyword arguments. See [here](https://github.com/unifyai/ivy/blob/e85a3f2b92b8768ad0c337937b4149f1385f4de6/ivy_tests/test_ivy/helpers/function_testing.py#L1645) for more examples of the nested functions in action.